### ISSUE TITLE
[SSR] Parent's state is always in initial state during server-side rendering. Application is rendered twice on the server

### ISSUE THREAD

**Issue by hinok** [apollographql/react-apollo#1285]
**Intended outcome:**
While server-side rendering, application is rendered including updated parent's state by resolved data from the child component that is wrapped with `graphql` HOC.

Example:

```jsx
<Parent>
  {(updateParentState) => (
    // This is component that is wrapped with graphql HOC
    <Child updateParentState={updateParentState} />
  )}
</Parent>
```

**Actual outcome:**
While server-side rendering, application is rendered but with initial parent's state, it looks like parent's state wasn't updated at all by the child component.

I'm using `renderToStringWithData` on the server.

**How to reproduce the issue:**

It's just an example based on my application code. I found this issue while building a general pagination HOC for my application.

**index.js**

```jsx
// <Pagination> doesn't know anything about graphql or fetching data
<Pagination>
  {(updateTotalItems, renderPagination) => (
    // Here <App> is wrapped with graphql() HOC
    <App updateTotalItems={updateTotalItems} renderPagination={renderPagination} />
  )}
</Pagination>
```

**Pagination.js**

```jsx
class Pagination extends Component {
  state = {
    totalItems: 0,
  };
  
  // logic...

  updateTotalItems = (totalItems) => {
    this.setState({ totalItems });
  };

  renderPagination = () => {
    const { totalItems } = this.state;
    const totalPages = Math.ceil(totalItems / numberOfItemsPerPage);

    // logic...
    // calculate total pages based on totalItems and number of items per page etc.
    // when everything is ready, just render component...

    return <PaginationLinks page={page} totalPages={totalPages} ... />;
  };

  render() {
    return this.props.children(
      this.updateTotalItems, 
      this.renderPagination
    );
  }
}
```

**App.js**

```jsx
class App extends Component {
  static propTypes = {
    updateTotalItems: PropTypes.func.isRequired,
    renderPagination: PropTypes.func.isRequired,
  };
  
  // This is called on the server and it set up parent's state (<Pagination /> component)
  componentWillMount() {
    if (!this.props.data.loading) {
      this.props.updateTotalItems(this.props.data.totalItems);
    }
  }
  
  componentWillReceiveProps(nextProps) {
    if (!nextProps.data.loading) {
      this.props.updateTotalItems(nextProps.data.totalItems);
    }
  }
  
  render() {
    <div>
      {/* render list of data... */}
      {this.props.renderPagination()}
    </div>
  }
}

export default graphql(MyQuery)(App);
```

1) Everything works correctly on the client-side.
2) While debugging, I found that `<Pagination>` state is correctly updated based on resolved data from the `<App>` component and even pagination is rendered, but while calling `renderToStringWithData`, it seems that application is rendered twice and during the 2nd render, `<Pagination>` state is in initial state for all the time.

```
// 1) Start rendering
renderToStringWithData is called

// 2) Looks like first render
App.componentWillMount() is called with resolved data
Pagination.updateTotalItems() is called with resolved data
Pagination state is updated
Pagination.renderPagination() is called and pagination links are rendered

// 3) Looks like second render
App.componentWillMount() is called with resolved data, AGAIN
Pagination.updateTotalItems() is called with resolved data
Pagination.renderPagination() is called with the initial state so links are not rendered because totalItems = 0 (state wasn't updated)

// 4) Return string representing application
HTML returned from renderToStringWithData doesn't contain pagination links...
```

Is this a bug? Has react-apollo some limitations in terms of server-side rendering .e.g. some things that works 100% correctly on the client-side, can stop working on the server-side like my example above?

**Version**
- apollo-client@1.9.3
- react-apollo@1.4.16
- react@15.6.1

**Comment by ravilution** [comment]
Hi,
I am facing similar issue. `renderToStringWithData` is trying to render the supplied React element twice. Please check the below code and output. I am using node v8.9.0

**package.json**
```javascript
{
  "name": "ssr-renders-twice",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "apollo-client-preset": "^1.0.3",
    "apollo-server-express": "^1.2.0",
    "express": "^4.16.2",
    "react": "^16.2.0",
    "react-apollo": "^2.0.1",
    "react-dom": "^16.2.0"
  }
}

```

**index.js**
```javascript
var React = require("React");
var express = require("express");
var renderToStringWithData = require("react-apollo").renderToStringWithData;

const app = express();

function App() {
  console.log("inside App");
  return React.createElement("div", null, "app");
}

app.get("/", async (req, res) => {
  console.log("req.url", req.url);
  const reactHTML = await renderToStringWithData(
    React.createElement(App, null)
  );
  res.send(`<!doctype html><html><body>${reactHTML}</body></html>`);
});

app.listen(3000, () => {
  console.log("started at ", new Date().toLocaleString());
});
```
**output**
```
started at  2017-11-28 14:07:57
req.url /
inside App
inside App
```

**Comment by LantosTG** [comment]
Same issue here,

reproduced at 
https://github.com/LantosTG/ReactIssueSSRDoubleCall

the issue is  in the renderToStringWithData function
export function renderToStringWithData(component) {
    return getDataFromTree(component).then(function () { return ReactDOM.renderToString(component); });
}

// the getDataFromTree will render the react component and then return the renderToString which renders again

**Comment by hwillson** [comment]
This should no longer be an issue with current day versions of `react-apollo`. Thanks!

**Comment by hinok** [comment]
@hwillson What version should fix this issue `3.0.0-beta.1` or `2.5.6`?
